### PR TITLE
IGNITE-18831 Sql. Fix dynamic parameters type coercion.

### PR DESCRIPTION
**Pull request by lowka**
Adds type family compatibility checks for dynamic parameters.

>  OK: SELECT * FROM t WHERE int_col = ?:int
>
>  OK: SELECT * FROM t WHERE int_col = ?:smallint
>         SELECT * FROM t WHERE int_col = cast(?:smallint, int)
>
>  ERR: SELECT * FROM t WHERE int_col = ?:str
>  ERR: SELECT * FROM t WHERE str_col = ?:int
>
>  ERR: SELECT 1 + ?:str
>  ERR: SELECT ?:str + ?:str

Adds type family compatibility checks for columns in set operations:

> OK: SELECT smallint_col FROM t1 UNION SELECT int_col FROM FROM t2
>  ERR: SELECT int_col FROM t1 UNION str_col FROM t2
> ERR: SELECT '1' UNION SELECT 2

Fixes coerced expression propagation in UPDATE/MERGE - https://issues.apache.org/jira/browse/IGNITE-19481

>UPDATE t SET str_col = int_lit is coerced to UPDATE t SET str_col = CAST(int_lit AS VARCHAR)
>but a CAST is lost and as a result we get cast cast exception during execution.

Introduces `StatementChecker` for check plans in unit tests - https://github.com/apache/ignite-3/pull/2047/files#diff-58eb8f3fdeedb3f2f8829aa6360e8b63fc058de965df79b8bc48743ab0fad640

Also makes type coercion in set operations illegal because at the moment type coercion for them is broken when a column on the other side of a set operator is a literal (https://issues.apache.org/jira/browse/CALCITE-5130) and there is no way to distinguish columns/literal/dynamic parameters.

### CODE REVIEW

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDynamicParameterTest.java [hunk 1]
```diff
@@ -164,15 +166,30 @@ public void testWithDifferentParametersTypes() {
      */
     @Test
     public void testWithDifferentParametersTypesMismatch() {
-        assertThrows(CalciteContextException.class, () -> assertQuery("SELECT COALESCE(12.2, ?)").withParams("b").returns(12.2).check());
-        assertThrows(CalciteContextException.class, () -> assertQuery("SELECT COALESCE(?, ?)").withParams(12.2, "b").returns(12.2).check());
+        assertThrows(CalciteContextException.class, () -> assertQuery("SELECT COALESCE(12.2, ?)").withParams("b").check());
```
Comment (zstan) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDynamicParameterTest.java:169]: why we no more check return result ?

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItFunctionsTest.java [hunk 2]
```diff
@@ -294,7 +294,6 @@ public void testSubstr() {
         assertQuery("SELECT SUBSTR('abcdefg', 2)").returns("bcdefg");
         assertQuery("SELECT SUBSTR('abcdefg', -1)").returns("abcdefg");
         assertQuery("SELECT SUBSTR('abcdefg', 1, -3)").returns("");
-        assertQuery("SELECT SUBSTR(1000, 1, 3)").returns("100");
```
Comment (zstan) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItFunctionsTest.java:297]: change it into : SELECT SUBSTR(1000::varchar(4), 1, 3)

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItImplicitCastsTest.java [hunk 3]
```diff
@@ -70,6 +77,8 @@ public void testMergeSort(ColumnPair columnPair) {
         sql(format("CREATE TABLE T11 (c1 int primary key, c2 {})", columnPair.lhs));
         sql(format("CREATE TABLE T12 (c1 int primary key, c2 {})", columnPair.rhs));
 
+        initData(columnPair);
```
Comment (zstan) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItImplicitCastsTest.java:80]: is it useful to add Integer - Double , Float - Double pair here ?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java [hunk 4]
```diff
@@ -557,82 +585,93 @@ private boolean isSystemFieldName(String alias) {
...
-                    }
-                }
-            }
+            RelDataType paramType = getDynamicParamType(param);
+            RelDataType derivedType = getValidatedNodeType(param);
 
-            for (int i = 0; i < operands.size(); ++i) {
-                SqlNode operand = operands.get(i);
+            // We can check for nullability, but it was set to true.
+            if (!SqlTypeUtil.equalSansNullability(derivedType, paramType)) {
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java:667]: you need to use org.apache.ignite.internal.util.IgniteUtils#assertionsEnabled or rework this approach at all

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java [hunk 5]
```diff
@@ -153,6 +160,48 @@ public void validateUpdate(SqlUpdate call) {
...
+    /** {@inheritDoc} */
+    @Override
+    public void validateMerge(SqlMerge call) {
+        super.validateMerge(call);
+
+        SqlSelect select = call.getSourceSelect();
+        SqlUpdate update = call.getUpdateCall();
+
+        if (update != null) {
+            assert select != null : "Merge: SourceSelect has not been set";
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java:181]: why we need only assertions on validation ?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/fun/IgniteSqlOperatorTable.java [hunk 6]
```diff
@@ -105,7 +105,7 @@ public class IgniteSqlOperatorTable extends ReflectiveSqlOperatorTable {
      */
     public static final SqlFunction SUBSTR = new SqlFunction("SUBSTR", SqlKind.OTHER_FUNCTION,
             ReturnTypes.ARG0_NULLABLE_VARYING, null,
-            OperandTypes.STRING_INTEGER_OPTIONAL_INTEGER,
+            OperandTypes.STRING_INTEGER.or(OperandTypes.STRING_INTEGER_INTEGER),
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/sql/fun/IgniteSqlOperatorTable.java:108]: whats the problem with previous implementation ?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java [hunk 7]
```diff
@@ -148,4 +149,37 @@ public RelDataType deriveSumType(RelDataTypeFactory typeFactory, RelDataType arg
 
         return typeFactory.createTypeWithNullability(sumType, argumentType.isNullable());
     }
+
+    /**
+     * Checks that {@code toType} and {@code fromType} have compatible type families taking into account custom data types.
+     *
+     * @see SqlTypeUtil#canAssignFrom(RelDataType, RelDataType)
+     */
+    public boolean typeFamiliesAreCompatible(RelDataTypeFactory typeFactory, RelDataType toType, RelDataType fromType) {
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java:158]: why you use keyword "family" here ?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java [hunk 8]
```diff
@@ -148,4 +149,37 @@ public RelDataType deriveSumType(RelDataTypeFactory typeFactory, RelDataType arg
...
+            return true;
+        } else if (fromType instanceof IgniteCustomType && toType instanceof IgniteCustomType) {
+            IgniteCustomType fromCustom = (IgniteCustomType) fromType;
+            IgniteCustomType toCustom = (IgniteCustomType) toType;
+
+            // IgniteCustomType: different custom data types are not compatible.
+            return Objects.equals(fromCustom.getCustomTypeName(), toCustom.getCustomTypeName());
+        } else if (fromType instanceof IgniteCustomType || toType instanceof IgniteCustomType) {
+            // Custom data types are not compatible with other types.
+            return false;
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java:177]: do we have such a tests?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java [hunk 9]
```diff
@@ -148,4 +149,37 @@ public RelDataType deriveSumType(RelDataTypeFactory typeFactory, RelDataType arg
...
+        } else {
+            boolean fromTo = SqlTypeUtil.canAssignFrom(toType, fromType);
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java:179]: whats the difference between 1 time call : SqlTypeUtil.inSameFamily(t1, t2) ?

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/util/StatementCheckerTest.java [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    public void testOk() throws Throwable {
+        when(sqlPrepare.prepare(any(IgniteSchema.class), any(String.class), any(List.class))).thenReturn(dummyNode);
+
+        DynamicTest test = newChecker().sql("SELECT 1").ok();
+        assertEquals("OK SELECT 1", test.getDisplayName(), "display name");
+
+        test.getExecutable().execute();
+    }
+
+    /** Validation check should pass - error any error is accepted. */
```
Comment (zstan) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/util/StatementCheckerTest.java:70]:
```suggestion
    /** Validation check should pass - any error is accepted. */
```

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteTypeCoercion.java [hunk 11]
```diff
@@ -113,6 +146,78 @@ public boolean binaryComparisonCoercion(SqlCallBinding binding) {
...
+        try {
+            return super.binaryArithmeticCoercion(binding);
+        } finally {
+            ctxStack.pop(ctx);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean builtinFunctionCoercion(SqlCallBinding binding, List<RelDataType> operandTypes, List<SqlTypeFamily> expectedFamilies) {
```
Comment (zstan) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteTypeCoercion.java:163]: seems this code have no tests ? isn`t it ?

**Review by AMashenkov — APPROVED** [review]

File: modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/datatypes/varbinary/ItVarBinaryQueryTest.java [hunk 12]
```diff
@@ -18,17 +18,38 @@
...
  * Tests for {@code SELECT} statement for {@link SqlTypeName#VARBINARY} type.
  */
 public class ItVarBinaryQueryTest extends BaseQueryDataTypeTest<VarBinary> {
 
+    /**
+     * {@code VARCHAR} vs type that can not be casted to {@code VARCHAR}.
+     */
+    @ParameterizedTest
+    @MethodSource("binaryComparisonOperators")
+    public void testInvalidComparisonOperation(String opSql, String opName) {
```
Comment (zstan) [inline modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/datatypes/varbinary/ItVarBinaryQueryTest.java:45]: opName is not used

**Comment by lowka** [comment]
Removed a link to resolved ticket.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java [hunk 13]
```diff
@@ -148,4 +149,36 @@ public RelDataType deriveSumType(RelDataTypeFactory typeFactory, RelDataType arg
 
         return typeFactory.createTypeWithNullability(sumType, argumentType.isNullable());
     }
+
+    /**
+     * Checks that {@code toType} and {@code fromType} have compatible type families taking into account custom data types.
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java:154]: Could you please clarify what compatibility means?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java [hunk 14]
```diff
@@ -148,4 +149,36 @@ public RelDataType deriveSumType(RelDataTypeFactory typeFactory, RelDataType arg
...
+    public boolean typeFamiliesAreCompatible(RelDataTypeFactory typeFactory, RelDataType toType, RelDataType fromType) {
+        // Types T1 and T2 have compatible type families if their types are assignable.
+
+        // Same types are always compatible.
+        if (SqlTypeUtil.equalSansNullability(typeFactory, toType, fromType)) {
+            return true;
+        }
+
+        // NULL is compatible with all types.
+        if (fromType.getSqlTypeName() == SqlTypeName.NULL || toType.getSqlTypeName() == SqlTypeName.NULL) {
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java:167]: > toType.getSqlTypeName() == SqlTypeName.NULL

Is it even possible?

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java [hunk 15]
```diff
@@ -148,4 +149,36 @@ public RelDataType deriveSumType(RelDataTypeFactory typeFactory, RelDataType arg
 
         return typeFactory.createTypeWithNullability(sumType, argumentType.isNullable());
     }
+
+    /**
+     * Checks that {@code toType} and {@code fromType} have compatible type families taking into account custom data types.
+     *
+     * @see SqlTypeUtil#canAssignFrom(RelDataType, RelDataType)
+     */
+    public boolean typeFamiliesAreCompatible(RelDataTypeFactory typeFactory, RelDataType toType, RelDataType fromType) {
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/type/IgniteTypeSystem.java:158]: to be honest, I don't understand this method. In my opinion, different real families are never compatible. One exception is for NULL, I guess

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java [hunk 16]
```diff
@@ -306,70 +358,46 @@ public void validateAggregateParams(SqlCall aggCall,
...
+        // Check compatibility for operands of binary comparison operation between custom data types vs built-in SQL types.
+        // We get here because in calcite ANY type can be assigned/casted to all other types.
+        // This check can be a part of some SqlOperandTypeChecker?
 
-        if (!canConvert) {
-            var ex = RESOURCE.invalidTypesForComparison(
-                    lhs.getFullTypeString(), sqlKind.sql, rhs.getFullTypeString());
+        if (lhs instanceof IgniteCustomType || rhs instanceof IgniteCustomType) {
+            boolean lhsRhsCompatible = IgniteTypeSystem.INSTANCE.typeFamiliesAreCompatible(typeFactory, lhs, rhs);
+            boolean rhsLhsCompatible = IgniteTypeSystem.INSTANCE.typeFamiliesAreCompatible(typeFactory, rhs, lhs);
```
Comment (korlov42) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java:392]: I believe the proper way to derive types is through type coercion

**Pull request merged**